// @serverless-stack/resources/dist/util/builder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = exports.getEsbuildMetafileName = void 0;
const chalk_1 = __importDefault(require("chalk"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const zip_local_1 = __importDefault(require("zip-local"));
const esbuild = __importStar(require("esbuild"));
const child_process_1 = require("child_process");
function addExtensionToHandler(handler, extension) {
    return handler.replace(/\.[\w\d]+$/, extension);
}
function getEsbuildMetafileName(handler) {
    const key = handler.replace(/[/.]/g, "-");
    return `.esbuild.${key}.json`;
}
exports.getEsbuildMetafileName = getEsbuildMetafileName;
function getEsbuildExternal(srcPath, bundle) {
    let externals = ["aws-sdk"];
    if (bundle) {
        return [
            ...externals,
            ...(bundle.externalModules || []),
            ...(bundle.nodeModules || []),
        ];
    }
    try {
        const packageJson = fs.readJsonSync(path.join(srcPath, "package.json"));
        externals = Object.keys(Object.assign(Object.assign(Object.assign({}, (packageJson.dependencies || {})), (packageJson.devDependencies || {})), (packageJson.peerDependencies || {})));
    }
    catch (e) {
        console.log(chalk_1.default.grey(`No package.json found in ${srcPath}`));
    }
    return externals;
}
function getEsbuildLoader(bundle) {
    if (bundle) {
        return bundle.loader || {};
    }
    return undefined;
}
function getHandlerFullPosixPath(srcPath, handler) {
    return srcPath === "." ? handler : `${srcPath}/${handler}`;
}
/**
 * Extract versions for a list of modules.
 *
 * First lookup the version in the package.json and then fallback to requiring
 * the module's package.json. The fallback is needed for transitive dependencies.
 */
function extractDependencies(pkgPath, modules) {
    var _a, _b, _c, _d;
    const dependencies = {};
    const pkgJson = fs.readJsonSync(pkgPath);
    const pkgDependencies = Object.assign(Object.assign(Object.assign({}, ((_a = pkgJson.dependencies) !== null && _a !== void 0 ? _a : {})), ((_b = pkgJson.devDependencies) !== null && _b !== void 0 ? _b : {})), ((_c = pkgJson.peerDependencies) !== null && _c !== void 0 ? _c : {}));
    for (const mod of modules) {
        try {
            const version = (_d = pkgDependencies[mod]) !== null && _d !== void 0 ? _d : require(`${mod}/package.json`).version; // eslint-disable-line @typescript-eslint/no-var-requires
            dependencies[mod] = version;
        }
        catch (err) {
            throw new Error(`Cannot extract version for module '${mod}'. Check that it's referenced in your package.json or installed.`);
        }
    }
    return dependencies;
}
function builder(builderProps) {
    const { target, bundle, srcPath, handler, buildDir } = builderProps;
    console.log(chalk_1.default.grey(`Building Lambda function ${getHandlerFullPosixPath(srcPath, handler)}`));
    // Check has tsconfig
    const tsconfig = path.join(srcPath, "tsconfig.json");
    const hasTsconfig = fs.existsSync(tsconfig);
    // Check entry path exists
    let entryPath = path.join(srcPath, addExtensionToHandler(handler, ".ts"));
    if (!fs.existsSync(entryPath)) {
        entryPath = path.join(srcPath, addExtensionToHandler(handler, ".js"));
    }
    if (!fs.existsSync(entryPath)) {
        throw new Error(`Cannot find a handler file at ${entryPath}".`);
    }
    // Four cases:
    //  1. BUNDLE + srcPath ROOT
    //      src       : path/to/file.method
    //      buildPath : .build/hash-$ts
    //      zipPath   : .build/hash-$ts
    //      outZip    : .build/hash-$ts.zip
    //      outHandler: file.method
    //
    //  2. BUNDLE + srcPath NON-ROOT
    //      src       : srcPath/path/to/file.method
    //      buildPath : srcPath/.build/hash-$ts
    //      zipPath   : srcPath/.build/hash-$ts
    //      outZip    : .build/hash-$ts.zip
    //      outHandler: file.method
    //
    //  3. non-BUNDLE + srcPath ROOT
    //      src       : path/to/file.method
    //      buildPath : .build/handlerDir
    //      zipPath   : .
    //
    //     Note: This case is NOT SUPPORTED because we need to zip the app root for each
    //           handler. So after a Lambda's zip is generated, the next Lambda's zip will
    //           contain the previous Lambda's zip inside .build, and the previous Lambda's
    //           zip inside cdk.out.
    //
    //           One solution would be to cherry pick what to zip. For example, zip should
    //           only include the esbuid's output (ie. .js and .js.map files) from the
    //           .build folder.
    //
    //           Also need to clear all .build folders generated from Lambda functions that
    //           has srcPath.
    //
    //  4. non-BUNDLE + srcPath NON-ROOT
    //      src       : srcPath/path/to/file.method
    //      buildPath : srcPath/.build/hash-$ts
    //      zipPath   : srcPath
    //      outZip    : .build/hash-$ts.zip
    //      outHandler: .build/hash-$ts/file.method
    //
    //     Note: place outZip at the app root's .build because entire srcPath is zipped up.
    //           If outZip is srcPath's .build, a Lambda's zip would include zip files from
    //           all the previous Lambdas.
    const appPath = process.cwd();
    const handlerHash = `${getHandlerFullPosixPath(srcPath, handler).replace(/[/.]/g, "-")}-${Date.now()}`;
    const buildPath = path.join(srcPath, buildDir, handlerHash);
    const metafile = path.join(srcPath, buildDir, getEsbuildMetafileName(handler));
    // Transpile
    transpile(entryPath);
    // Package nodeModules
    installNodeModules(srcPath, bundle);
    // Copy files
    copyFiles(bundle);
    // Zip
    let outZip, outHandler;
    if (bundle) {
        outZip = path.join(appPath, buildDir, `${handlerHash}.zip`);
        outHandler = path.basename(handler);
        zip(buildPath, outZip);
    }
    else {
        outZip = path.join(appPath, buildDir, `${handlerHash}.zip`);
        outHandler = `${buildDir}/${handlerHash}/${path.basename(handler)}`;
        zip(srcPath, outZip);
    }
    return { outZip, outHandler };
    function transpile(entryPath) {
        esbuild.buildSync({
            external: getEsbuildExternal(srcPath, bundle),
            loader: getEsbuildLoader(bundle),
            metafile,
            bundle: true,
            format: "cjs",
            sourcemap: true,
            platform: "node",
            target: [target],
            outdir: buildPath,
            entryPoints: [entryPath],
            color: process.env.NO_COLOR !== "true",
            tsconfig: hasTsconfig ? tsconfig : undefined,
        });
    }
    function installNodeModules(srcPath, bundle) {
        // Validate 'nodeModules' is defined in bundle options
        bundle = bundle;
        if (!bundle || !bundle.nodeModules || bundle.nodeModules.length === 0) {
            return;
        }
        // Find 'package.json' at handler's srcPath.
        const pkgPath = path.join(srcPath, "package.json");
        if (!fs.existsSync(pkgPath)) {
            throw new Error(`Cannot find a "package.json" in the function's srcPath: ${path.resolve(srcPath)}`);
        }
        // Determine dependencies versions, lock file and installer
        const dependencies = extractDependencies(pkgPath, bundle.nodeModules);
        let installer = "npm";
        let lockFile;
        if (fs.existsSync(path.join(srcPath, "package-lock.json"))) {
            installer = "npm";
            lockFile = "package-lock.json";
        }
        else if (fs.existsSync(path.join(srcPath, "yarn.lock"))) {
            installer = "yarn";
            lockFile = "yarn.lock";
        }
        // Create dummy package.json, copy lock file if any and then install
        const outputPath = path.join(buildPath, "package.json");
        fs.ensureFileSync(outputPath);
        fs.writeJsonSync(outputPath, { dependencies });
        if (lockFile) {
            fs.copySync(path.join(srcPath, lockFile), path.join(buildPath, lockFile));
        }
        try {
            child_process_1.execSync(`${installer} install`, {
                cwd: buildPath,
                stdio: "pipe",
            });
        }
        catch (e) {
            console.log(e.stdout.toString());
            console.log(e.stderr.toString());
            throw new Error("There was a problem installing nodeModules.");
        }
    }
    function copyFiles(bundle) {
        // Validate 'copyFiles' is defined in bundle options
        bundle = bundle;
        if (!bundle || !bundle.copyFiles || bundle.copyFiles.length === 0) {
            return;
        }
        bundle.copyFiles.forEach(({ from, to }) => {
            const fromPath = path.join(srcPath, from);
            const toPath = path.join(buildPath, to);
            fs.copySync(fromPath, toPath);
        });
    }
    function zip(dir, zipFile) {
        try {
            zip_local_1.default.sync.zip(dir).compress().save(zipFile);
        }
        catch (e) {
            console.log(e);
            throw new Error("There was a problem generating Lambda package.");
        }
        return zipFile;
    }
}
exports.builder = builder;
