// projen/lib/github/auto-merge.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoMerge = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("../component");
/**
 * (experimental) Sets up mergify to merging approved pull requests.
 *
 * If `buildJob` is specified, the specified GitHub workflow job ID is required
 * to succeed in order for the PR to be merged.
 *
 * `approvedReviews` specified the number of code review approvals required for
 * the PR to be merged.
 *
 * @experimental
 */
class AutoMerge extends component_1.Component {
    /**
     * @experimental
     */
    constructor(project, options = {}) {
        var _b, _c, _d, _e;
        super(project);
        const successfulBuild = options.buildJob
            ? [`status-success=${options.buildJob}`]
            : [];
        const mergeAction = {
            merge: {
                // squash all commits into a single commit when merging
                method: 'squash',
                // use PR title+body as the commit message
                commit_message: 'title+body',
                // update PR branch so it's up-to-date before merging
                strict: 'smart',
                strict_method: 'merge',
            },
            delete_head_branch: {},
        };
        const approvedReviews = (_b = options.approvedReviews) !== null && _b !== void 0 ? _b : 1;
        (_c = project.github) === null || _c === void 0 ? void 0 : _c.addMergifyRules({
            name: 'Automatic merge on approval and successful build',
            actions: mergeAction,
            conditions: [
                `#approved-reviews-by>=${approvedReviews}`,
                ...successfulBuild,
            ],
        });
        // empty string means disabled.
        const autoMergeLabel = (_d = options.autoMergeLabel) !== null && _d !== void 0 ? _d : 'auto-merge';
        ;
        this.autoMergeLabel = autoMergeLabel;
        if (this.autoMergeLabel) {
            (_e = project.github) === null || _e === void 0 ? void 0 : _e.addMergifyRules({
                name: `Automatic merge PRs with ${autoMergeLabel} label upon successful build`,
                actions: mergeAction,
                conditions: [
                    `label=${autoMergeLabel}`,
                    ...successfulBuild,
                ],
            });
        }
    }
}
exports.AutoMerge = AutoMerge;
_a = JSII_RTTI_SYMBOL_1;
AutoMerge[_a] = { fqn: "projen.github.AutoMerge", version: "0.17.1" };
