// projen/lib/cleanup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cleanup = void 0;
const path = require("path");
const fs = require("fs-extra");
const glob = require("glob");
const common_1 = require("./common");
const logging = require("./logging");
function cleanup(dir, exclude) {
    try {
        for (const f of findGeneratedFiles(dir, exclude)) {
            fs.removeSync(f);
        }
    }
    catch (e) {
        logging.warn(`warning: failed to clean up generated files: ${e.stack}`);
    }
}
exports.cleanup = cleanup;
function findGeneratedFiles(dir, exclude) {
    const ignore = [...readGitIgnore(dir), 'node_modules/**', ...exclude, '.git/**'];
    const files = glob.sync('**', {
        ignore,
        cwd: dir,
        dot: true,
        nodir: true,
        absolute: true,
    });
    const generated = new Array();
    for (const file of files) {
        const contents = fs.readFileSync(file, 'utf-8');
        if (contents.includes(common_1.PROJEN_MARKER)) {
            generated.push(file);
        }
    }
    return generated;
}
function readGitIgnore(dir) {
    const filepath = path.join(dir, '.gitignore');
    if (!fs.pathExistsSync(filepath)) {
        return [];
    }
    return fs.readFileSync(filepath, 'utf-8')
        .split('\n')
        .filter(x => (x === null || x === void 0 ? void 0 : x.trim()) !== '')
        .filter(x => !x.startsWith('#') && !x.startsWith('!'))
        .map(x => x.replace(/^\//, '')) // remove "/" prefix
        .map(x => `${x}\n${x}/**`)
        .join('\n')
        .split('\n');
}
