// projen/lib/ignore-file.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IgnoreFile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const file_1 = require("./file");
/**
 * @experimental
 */
class IgnoreFile extends file_1.FileBase {
    /**
     * @experimental
     */
    constructor(project, filePath) {
        super(project, filePath, { editGitignore: filePath !== '.gitignore' });
        this._excludes = new Set();
        this._includes = new Set();
    }
    /**
     * (experimental) Add ignore patterns.
     *
     * Files that match this pattern will be ignored. If the
     * pattern starts with a negation mark `!`, files that match will _not_ be
     * ignored.
     *
     * Comment lines (start with `#`) are ignored.
     *
     * @param patterns Ignore patterns.
     * @experimental
     */
    addPatterns(...patterns) {
        for (const pattern of patterns) {
            // skip comments
            if (pattern.startsWith('#')) {
                continue;
            }
            if (pattern.startsWith('!')) {
                this._includes.add(pattern);
            }
            else {
                this._excludes.add(pattern);
            }
        }
    }
    /**
     * (experimental) Removes patterns previously added from the ignore file.
     *
     * If `addPattern()` is called after this, the pattern will be added again.
     *
     * @param patterns patters to remove.
     * @experimental
     */
    removePatterns(...patterns) {
        for (const p of patterns) {
            this._excludes.delete(p);
            this._includes.delete(p);
        }
    }
    /**
     * (experimental) Ignore the files that match these patterns.
     *
     * @param patterns The patterns to match.
     * @experimental
     */
    exclude(...patterns) {
        return this.addPatterns(...patterns);
    }
    /**
     * (experimental) Always include the specified file patterns.
     *
     * @param patterns Patterns to include in git commits.
     * @experimental
     */
    include(...patterns) {
        for (let pattern of patterns) {
            if (!pattern.startsWith('!')) {
                pattern = '!' + pattern;
            }
            this.addPatterns(pattern);
        }
    }
    /**
     * (experimental) Implemented by derived classes and returns the contents of the file to emit.
     *
     * @experimental
     */
    synthesizeContent(resolver) {
        const lines = [`# ${file_1.FileBase.PROJEN_MARKER}`];
        for (const line of Array.from(this._excludes).sort()) {
            // if this line exists in the inclusion list, then skip it
            if (this._includes.has(`!${line}`)) {
                continue;
            }
            lines.push(line);
        }
        for (const line of Array.from(this._includes).sort()) {
            lines.push(line);
        }
        return `${resolver.resolve(lines).join('\n')}\n`;
    }
}
exports.IgnoreFile = IgnoreFile;
_a = JSII_RTTI_SYMBOL_1;
IgnoreFile[_a] = { fqn: "projen.IgnoreFile", version: "0.17.1" };
