// projen/lib/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disable = exports.warn = exports.error = exports.info = exports.verbose = exports.debug = void 0;
const chalk = require("chalk");
let enabled = true;
function log(color, ...text) {
    if (!enabled) {
        return;
    }
    console.error(`🤖 ${color(...text)}`);
}
/**
 * @deprecated
 */
function debug(...text) {
    if (process.env.DEBUG) {
        log(chalk.gray, ...text);
    }
}
exports.debug = debug;
/**
 * @deprecated
 */
function verbose(...text) {
    log(chalk.white, ...text);
}
exports.verbose = verbose;
/**
 * @deprecated
 */
function info(...text) {
    log(chalk.cyan, ...text);
}
exports.info = info;
/**
 * @deprecated
 */
function error(...text) {
    log(chalk.red, ...text);
}
exports.error = error;
/**
 * @deprecated
 */
function warn(...text) {
    log(chalk.yellow, ...text);
}
exports.warn = warn;
/**
 * @deprecated
 */
function disable() {
    enabled = false;
}
exports.disable = disable;
