// projen/lib/makefile.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Makefile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const file_1 = require("./file");
/**
 * (experimental) Minimal Makefile.
 *
 * @experimental
 */
class Makefile extends file_1.FileBase {
    /**
     * @experimental
     */
    constructor(project, filePath, options = {}) {
        super(project, filePath, options);
        const all = options.all ? options.all : [];
        const rules = options.rules ? options.rules : [];
        rules.forEach(e => this.validateRule(e));
        this.all = {
            targets: ['all'],
            prerequisites: all,
            phony: true,
        };
        this.rules = [
            this.all,
            ...rules,
        ];
    }
    /**
     * (experimental) Add a target to all.
     *
     * @experimental
     */
    addAll(target) {
        this.all.prerequisites.push(target);
        return this;
    }
    /**
     * (experimental) Add multiple targets to all.
     *
     * @experimental
     */
    addAlls(...targets) {
        targets.forEach(e => this.addAll(e));
        return this;
    }
    /**
     * (experimental) Add a rule to the Makefile.
     *
     * @experimental
     */
    addRule(rule) {
        this.validateRule(rule);
        this.rules.push(rule);
        return this;
    }
    /**
     * (experimental) Add multiple rules to the Makefile.
     *
     * @experimental
     */
    addRules(...rules) {
        rules.forEach(e => this.addRule(e));
        return this;
    }
    validateRule(rule) {
        if (!rule.targets || !rule.targets.length) {
            throw new Error('"targets" cannot be undefined or empty for items in "rules"');
        }
    }
    /**
     * (experimental) Implemented by derived classes and returns the contents of the file to emit.
     *
     * @experimental
     */
    synthesizeContent(resolver) {
        const rules = resolver.resolve(this.rules);
        const lines = [
            ...rules.map((rule) => {
                const targets = rule.targets.join(' ');
                const prerequisites = (rule.prerequisites ? rule.prerequisites : []).join(' ');
                const recipe = rule.recipe ? rule.recipe : [];
                const phony = rule.phony ? [`.PHONY: ${targets}`] : [];
                return [
                    ...phony,
                    `${targets}: ${prerequisites}`.trim(),
                    ...recipe.map(step => `\t${step}`),
                ].join('\n');
            }),
        ];
        return `${lines.join('\n\n')}\n`;
    }
}
exports.Makefile = Makefile;
_a = JSII_RTTI_SYMBOL_1;
Makefile[_a] = { fqn: "projen.Makefile", version: "0.17.1" };
