// projen/lib/projen-upgrade.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjenUpgrade = void 0;
const tasks_1 = require("./tasks");
/**
 * Checks for new versions of projen and creates a PR with an upgrade change.
 */
class ProjenUpgrade {
    constructor(project, options = {}) {
        var _a;
        const upgradeTask = project.addTask('projen:upgrade', {
            description: 'upgrades projen to the latest version',
            category: tasks_1.TaskCategory.MAINTAIN,
        });
        upgradeTask.exec('yarn upgrade -L projen');
        upgradeTask.exec('CI="" yarn projen');
        if (options.autoUpgradeSecret) {
            if (!project.github) {
                throw new Error('github workflows are required in order for auto-update');
            }
            const workflow = project.github.addWorkflow('ProjenUpgrade');
            workflow.on({
                schedule: options.autoUpgradeSchedule
                    ? options.autoUpgradeSchedule.map(s => ({ cron: s }))
                    : [{ cron: '0 6 * * *' }],
                workflow_dispatch: {},
            });
            const withOptions = {
                'token': '${{ secrets.' + options.autoUpgradeSecret + ' }}',
                'commit-message': 'chore: upgrade projen',
                'branch': 'auto/projen-upgrade',
                'title': 'chore: upgrade projen',
                'body': 'This PR upgrades projen to the latest version',
            };
            if ((_a = options.labels) === null || _a === void 0 ? void 0 : _a.length) {
                withOptions.labels = options.labels.join(',');
            }
            workflow.addJobs({
                upgrade: {
                    'runs-on': 'ubuntu-latest',
                    'steps': [
                        // check out sources.
                        {
                            name: 'Checkout',
                            uses: 'actions/checkout@v2',
                        },
                        // install dependencies (and runs projen)
                        ...project.installWorkflowSteps,
                        // upgrade
                        {
                            name: 'Upgrade projen',
                            run: project.runTaskCommand(upgradeTask),
                        },
                        // submit a PR
                        {
                            name: 'Create pull request',
                            uses: 'peter-evans/create-pull-request@v3',
                            with: withOptions,
                        },
                    ],
                },
            });
        }
    }
}
exports.ProjenUpgrade = ProjenUpgrade;
