// projen/lib/version.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Version = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const path = require("path");
const fs = require("fs-extra");
const component_1 = require("./component");
const json_1 = require("./json");
const tasks_1 = require("./tasks");
const VERSION_FILE = 'version.json';
/**
 * @experimental
 */
class Version extends component_1.Component {
    /**
     * @experimental
     */
    constructor(project, options) {
        var _b;
        super(project);
        // this command determines if there were any changes since the last release
        // (the top-most commit is not a bump). it is used as a condition for both
        // the `bump` and the `release` tasks.
        const changesSinceLastRelease = '! git log --oneline -1 | grep -q "chore(release):"';
        this.bumpTask = project.addTask('bump', {
            description: 'Commits a bump to the package version based on conventional commits',
            category: tasks_1.TaskCategory.RELEASE,
            exec: 'standard-version',
            condition: changesSinceLastRelease,
        });
        const release = project.addTask('release', {
            description: `Bumps version & push to ${options.releaseBranch}`,
            category: tasks_1.TaskCategory.RELEASE,
            condition: changesSinceLastRelease,
        });
        release.spawn(this.bumpTask);
        release.exec(`git push --follow-tags origin ${options.releaseBranch}`);
        project.addDevDeps('standard-version@^9');
        (_b = project.npmignore) === null || _b === void 0 ? void 0 : _b.exclude('/.versionrc.json');
        project.gitignore.include(VERSION_FILE);
        let projenCommand = project.package.projenCommand;
        if (project.parent) {
            projenCommand = `cd ${path.relative(project.outdir, project.root.outdir)} && ${project.package.projenCommand}`;
        }
        new json_1.JsonFile(project, '.versionrc.json', {
            obj: {
                packageFiles: [{ filename: VERSION_FILE, type: 'json' }],
                bumpFiles: [{ filename: VERSION_FILE, type: 'json' }],
                commitAll: true,
                scripts: {
                    // run projen after release to update package.json
                    postbump: `${projenCommand} && git add .`,
                },
            },
        });
    }
    /**
     * (experimental) Returns the current version of the project.
     *
     * @experimental
     */
    get currentVersion() {
        const outdir = this.project.outdir;
        const versionFile = `${outdir}/${VERSION_FILE}`;
        if (!fs.existsSync(versionFile)) {
            if (!fs.existsSync(outdir)) {
                fs.mkdirpSync(outdir);
            }
            fs.writeFileSync(versionFile, JSON.stringify({ version: '0.0.0' }));
        }
        return JSON.parse(fs.readFileSync(versionFile, 'utf-8')).version;
    }
}
exports.Version = Version;
_a = JSII_RTTI_SYMBOL_1;
Version[_a] = { fqn: "projen.Version", version: "0.17.1" };
