// projen/lib/__tests__/python/python-project.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("../../logger");
const python_1 = require("../../python");
const util_1 = require("../util");
test('defaults', () => {
    const p = new TestPythonProject();
    expect(util_1.synthSnapshot(p)).toMatchSnapshot();
});
test('dependencies', () => {
    const p = new TestPythonProject();
    p.addDependency('Django@3.1.5');
    p.addDependency('aws-cdk.core@*');
    p.addDevDependency('hypothesis@^6.0.3');
    expect(util_1.synthSnapshot(p)).toMatchSnapshot();
});
test('dependencies via ctor', () => {
    const p = new TestPythonProject({
        deps: [
            'Django@3.1.5',
            'aws-cdk.core',
        ],
        devDeps: [
            'hypothesis@^6.0.3',
        ],
    });
    expect(util_1.synthSnapshot(p)).toMatchSnapshot();
});
test('no pytest', () => {
    const p = new TestPythonProject({
        pytest: false,
    });
    expect(util_1.synthSnapshot(p)).toMatchSnapshot();
});
test('pytest maxfailures', () => {
    const p = new TestPythonProject({
        pytestOptions: {
            maxFailures: 3,
        },
    });
    expect(util_1.synthSnapshot(p)['.projen/tasks.json'].tasks.test.steps[0].exec).toContain('--maxfail=3');
});
class TestPythonProject extends python_1.PythonProject {
    constructor(options = {}) {
        super({
            ...options,
            clobber: false,
            name: 'test-python-project',
            moduleName: 'test_python_project',
            authorName: 'First Last',
            authorEmail: 'email@example.com',
            version: '0.1.0',
            outdir: util_1.mkdtemp(),
            logging: { level: logger_1.LogLevel.OFF },
            jsiiFqn: 'projen.python.PythonProject',
        });
    }
}
