// projen/lib/__tests__/python/setuptools.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("../../logger");
const python_1 = require("../../python");
const util_1 = require("../util");
test('setuptools enabled', () => {
    const p = new TestPythonProject({
        setuptools: true,
        homepage: 'http://www.example.com',
        description: 'a short project description',
        license: 'Apache Software License',
        classifiers: [
            'Development Status :: 4 - Beta',
        ],
    });
    const snapshot = util_1.synthSnapshot(p);
    expect(snapshot['setup.py']).toContain('First Last');
    expect(snapshot['setup.py']).toContain('email@example.com');
    expect(snapshot['setup.py']).toContain('http://www.example.com');
    expect(snapshot['setup.py']).toContain('a short project description');
    expect(snapshot['setup.py']).toContain('Apache Software License');
    expect(snapshot['setup.py']).toContain('Development Status :: 4 - Beta');
});
class TestPythonProject extends python_1.PythonProject {
    constructor(options = {}) {
        super({
            ...options,
            clobber: false,
            name: 'test-python-project',
            moduleName: 'test_python_project',
            authorName: 'First Last',
            authorEmail: 'email@example.com',
            version: '0.1.0',
            outdir: util_1.mkdtemp(),
            logging: { level: logger_1.LogLevel.OFF },
            jsiiFqn: 'projen.python.PythonProject',
        });
    }
}
