// projen/lib/cli/cmds/start-app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.printStartMenu = exports.showStartMenu = void 0;
const chalk = require("chalk");
const inquirer = require("inquirer");
const tasks_1 = require("../../tasks");
const EXIT_MARKER = '$exit';
async function showStartMenu(tasks) {
    const { command } = await inquirer.prompt([
        {
            type: 'list',
            name: 'command',
            message: 'Scripts:',
            choices: renderChoices(tasks),
            pageSize: 100,
            loop: false,
        },
    ]);
    if (command === EXIT_MARKER) {
        return;
    }
    tasks.runTask(command);
}
exports.showStartMenu = showStartMenu;
function printStartMenu(tasks, root) {
    if (root && root !== '.') {
        console.error(chalk.cyanBright.bold(`Project: ${root}`));
    }
    console.error(chalk.cyanBright.underline('Commands:'));
    for (const entry of renderChoices(tasks)) {
        if (entry.type === 'separator') {
            console.error(entry.line);
        }
        else if (entry.name && entry.value !== '$exit') {
            console.error(entry.name);
        }
    }
}
exports.printStartMenu = printStartMenu;
function renderChoices(tasksrt) {
    var _a, _b;
    const tasks = tasksrt.tasks;
    const taskNames = tasks.map(t => t.name);
    const result = new Array();
    let category;
    const width = Math.max(...taskNames.map(k => k.length));
    for (const task of tasks.sort(sortByPriority)) {
        const cat = (_a = task.category) !== null && _a !== void 0 ? _a : tasks_1.TaskCategory.MISC;
        if (cat !== category) {
            result.push(new inquirer.Separator('  '));
            result.push(new inquirer.Separator(headingForCategory(cat)));
        }
        category = cat;
        result.push({
            name: `${task.name.padEnd(width)}   ${(_b = task.description) !== null && _b !== void 0 ? _b : ''}`,
            value: task.name,
            short: task.description,
        });
    }
    result.push(new inquirer.Separator('  '));
    result.push({
        name: 'EXIT',
        value: EXIT_MARKER,
    });
    return result;
}
function headingForCategory(category) {
    switch (category) {
        case tasks_1.TaskCategory.BUILD: return 'BUILD';
        case tasks_1.TaskCategory.TEST: return 'TEST';
        case tasks_1.TaskCategory.RELEASE: return 'RELEASE';
        case tasks_1.TaskCategory.MAINTAIN: return 'MAINTAIN';
        case tasks_1.TaskCategory.MISC:
        default:
            return 'MISC';
    }
}
function sortByPriority(e1, e2) {
    var _a, _b;
    const p1 = (_a = e1.category) !== null && _a !== void 0 ? _a : tasks_1.TaskCategory.MISC;
    const p2 = (_b = e2.category) !== null && _b !== void 0 ? _b : tasks_1.TaskCategory.MISC;
    if (p1 > p2)
        return 1;
    if (p1 < p2)
        return -1;
    return 0;
}
