// projen/lib/__tests__/java/java-project.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = require("fs-extra");
const java_project_1 = require("../../java/java-project");
const logger_1 = require("../../logger");
const util_1 = require("../util");
let cwd = process.cwd();
beforeEach(() => process.chdir(util_1.mkdtemp()));
afterEach(() => process.chdir(cwd));
test('defaults', () => {
    const p = new TestJavaProject();
    expect(util_1.synthSnapshot(p)).toMatchSnapshot();
});
test('pom options', () => {
    const p = new TestJavaProject({
        groupId: 'com.myorg',
        artifactId: 'play-202101050157',
    });
    snapPom(p);
});
test('dependencies', () => {
    const p = new TestJavaProject();
    p.pom.addDependency('software.amazon.awscdk/core@^1.2.3');
    p.pom.addTestDependency('org.assertj/assertj-core@3.18.1');
    p.pom.addPlugin('org.apache.maven.plugins/maven-compiler-plugin@3.8.1', {
        configuration: {
            source: '1.8',
            target: '1.8',
        },
    });
    snapPom(p);
});
test('dependencies via ctor', () => {
    const p = new TestJavaProject({
        deps: [
            'software.amazon.awscdk/core@^1.2.3',
            'software.amazon.awscdk/aws-s3@^1',
        ],
        testDeps: [
            'org.assertj/assertj-core@^3',
        ],
    });
    snapPom(p);
});
test('no junit', () => {
    const p = new TestJavaProject({
        junit: false,
    });
    expect(util_1.synthSnapshot(p)).toMatchSnapshot();
});
test('projenrc in java', () => {
    const p = new TestJavaProject();
    expect(util_1.synthSnapshot(p)['src/test/java/projenrc.java']).toMatchSnapshot();
});
test('disable projenrc in java', () => {
    const p = new TestJavaProject({
        projenrcJava: false,
    });
    expect(util_1.synthSnapshot(p)['src/test/java/projenrc.java']).toBeUndefined();
});
function snapPom(p) {
    expect(util_1.synthSnapshot(p)['pom.xml']).toMatchSnapshot();
}
class TestJavaProject extends java_project_1.JavaProject {
    constructor(options = {}) {
        // using a subdirectory to ensure synthSnapshot can clean up the project safely
        fs_extra_1.mkdirpSync('project-dir');
        super({
            ...options,
            // not using outdir: mkdtemp() since that will make snapshots non-deterministic
            outdir: './project-dir',
            groupId: 'org.acme',
            artifactId: 'my-artifact',
            name: 'test-project',
            version: '1.0.0',
            logging: { level: logger_1.LogLevel.OFF },
            jsiiFqn: 'projen.java.JavaProject',
            projenrcJavaOptions: {
                projenVersion: '^1.2.3',
            },
        });
    }
}
