// projen/lib/__tests__/java/pom.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const deps_1 = require("../../deps");
const java_1 = require("../../java");
const util_1 = require("../util");
test('group/artifact/version', () => {
    const pom = new TestPom({
        groupId: 'my.group.id',
        artifactId: 'mfoo-bar',
        version: '1.2.3',
    });
    expect(actualPom(pom)).toMatchSnapshot();
});
test('optional metadata', () => {
    const pom = new TestPom({
        groupId: 'my.group.id',
        artifactId: 'mfoo-bar',
        version: '1.2.3',
        description: 'hello, world!',
        url: 'https://foo/bar',
        packaging: 'not_jar',
    });
    expect(actualPom(pom)).toMatchSnapshot();
});
test('addProperty()', () => {
    const pom = new TestPom();
    pom.addProperty('project.build.sourceEncoding', 'UTF-8');
    pom.addProperty('junit.version', '5.7.0');
    expect(actualPom(pom)).toMatchSnapshot();
});
test('addDependency()', () => {
    const pom = new TestPom();
    pom.addDependency('software.amazon.awscdk/core');
    // dependencies are managed at the project level
    pom.project.deps.addDependency('org.assertj/assertj-core@^3.18.1', deps_1.DependencyType.TEST);
    expect(actualPom(pom)).toMatchSnapshot();
    // check that pom.addDependency() updates the project
    expect(pom.project.deps.getDependency('software.amazon.awscdk/core')).toStrictEqual({
        name: 'software.amazon.awscdk/core',
        type: 'runtime',
    });
});
test('addPlugin()', () => {
    const pom = new TestPom();
    pom.addPlugin('org.apache.maven.plugins/maven-compiler-plugin@3.8.1', {
        dependencies: [
            'org.projen/projen@^0.14',
        ],
        configuration: {
            source: '1.8',
            target: '1.8',
        },
    });
    // alteratively
    pom.project.deps.addDependency('org.codehaus.mojo/exec-maven-plugin@3.0.0', deps_1.DependencyType.BUILD, {
        configuration: {
            mainClass: 'com.myorg.Play202101050157App',
        },
    });
    expect(actualPom(pom)).toMatchSnapshot();
});
function actualPom(p) {
    const snap = util_1.synthSnapshot(p.project);
    return snap[p.fileName];
}
class TestPom extends java_1.Pom {
    constructor(options) {
        super(new util_1.TestProject(), options !== null && options !== void 0 ? options : {
            groupId: 'org.acme',
            artifactId: 'my-artifact',
            version: '1.2.3',
        });
    }
}
