// projen/lib/__tests__/java/projenrc.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const java_1 = require("../../java");
const projenrc_1 = require("../../java/projenrc");
const util_1 = require("../util");
test('projenrc.java support', () => {
    // GIVEN
    const project = new util_1.TestProject();
    const pom = new java_1.Pom(project, {
        groupId: 'my.group.id',
        artifactId: 'hello-world',
        version: '1.2.3',
    });
    // WHEN
    new projenrc_1.Projenrc(project, pom, {
        projenVersion: '1.0.0',
    });
    // THEN
    expect(util_1.synthSnapshot(project)).toMatchSnapshot();
});
test('use a non-test scope', () => {
    // GIVEN
    const project = new util_1.TestProject();
    const pom = new java_1.Pom(project, {
        groupId: 'my.group.id',
        artifactId: 'hello-world',
        version: '1.2.3',
    });
    // WHEN
    new projenrc_1.Projenrc(project, pom, {
        testScope: false,
        projenVersion: '7.7.7',
    });
    // THEN
    expect(util_1.synthSnapshot(project)).toMatchSnapshot();
});
test('set the class name', () => {
    // GIVEN
    const project = new util_1.TestProject();
    const pom = new java_1.Pom(project, {
        groupId: 'my.group.id',
        artifactId: 'hello-world',
        version: '1.2.3',
    });
    // WHEN
    new projenrc_1.Projenrc(project, pom, {
        className: 'boom.bam.projenrc',
        projenVersion: '7.7.7',
    });
    // THEN
    expect(util_1.synthSnapshot(project)).toMatchSnapshot();
});
test('generate projenrc in java', () => {
    // GIVEN
    const project = new util_1.TestProject({
        jsiiFqn: 'projen.java.JavaProject',
    });
    const pom = new java_1.Pom(project, {
        groupId: 'my.group.id',
        artifactId: 'hello-world',
        version: '1.2.3',
    });
    // WHEN
    new projenrc_1.Projenrc(project, pom, {
        initializationOptions: {
            junit: false,
        },
    });
    // THEN
    expect(util_1.synthSnapshot(project)['src/test/java/projenrc.java']).toMatchSnapshot();
});
