// projen/lib/java/junit.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Junit = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const path_1 = require("path");
const component_1 = require("../component");
const sample_file_1 = require("../sample-file");
const tasks_1 = require("../tasks");
const TESTDIR = path_1.join('src', 'test', 'java');
/**
 * (experimental) Implements JUnit-based testing.
 *
 * @experimental
 */
class Junit extends component_1.Component {
    /**
     * @experimental
     */
    constructor(project, options) {
        var _b, _c;
        super(project);
        const pom = options.pom;
        const version = (_b = options.version) !== null && _b !== void 0 ? _b : '5.7.0';
        pom.addTestDependency(`org.junit.jupiter/junit-jupiter-api@${version}`);
        pom.addTestDependency(`org.junit.jupiter/junit-jupiter-engine@${version}`);
        project.addTask('test', {
            description: 'Runs tests',
            category: tasks_1.TaskCategory.TEST,
            exec: 'mvn test',
        });
        const javaPackage = (_c = options.sampleJavaPackage) !== null && _c !== void 0 ? _c : 'org.acme';
        const javaPackagePath = javaPackage.split('.');
        new sample_file_1.SampleDir(project, path_1.join(TESTDIR, ...javaPackagePath), {
            files: {
                'MyTest.java': [
                    `package ${javaPackage};`,
                    '',
                    'import org.junit.jupiter.api.Test;',
                    '',
                    'public class MyTest {',
                    '  @Test',
                    '  public void testHello() {',
                    '    System.out.println("Hello, world!");',
                    '  }',
                    '}',
                ].join('\n'),
            },
        });
    }
}
exports.Junit = Junit;
_a = JSII_RTTI_SYMBOL_1;
Junit[_a] = { fqn: "projen.java.Junit", version: "0.17.1" };
