// projen/lib/java/maven-compile.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MavenCompile = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("../component");
const tasks_1 = require("../tasks");
/**
 * (experimental) Adds the maven-compiler plugin to a POM file and the `compile` task.
 *
 * @experimental
 */
class MavenCompile extends component_1.Component {
    /**
     * @experimental
     */
    constructor(project, pom, options = {}) {
        var _b, _c;
        super(project);
        project.gitignore.exclude('target');
        pom.addPlugin('org.apache.maven.plugins/maven-compiler-plugin@3.8.1', {
            configuration: {
                source: (_b = options.source) !== null && _b !== void 0 ? _b : '1.8',
                target: (_c = options.target) !== null && _c !== void 0 ? _c : '1.8',
            },
        });
        this.compileTask = project.addTask('compile', {
            description: 'Compile the main source files',
            category: tasks_1.TaskCategory.BUILD,
            exec: 'mvn compiler:compile',
        });
    }
}
exports.MavenCompile = MavenCompile;
_a = JSII_RTTI_SYMBOL_1;
MavenCompile[_a] = { fqn: "projen.java.MavenCompile", version: "0.17.1" };
