// projen/lib/java/maven-packaging.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MavenPackaging = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("../component");
const tasks_1 = require("../tasks");
/**
 * (experimental) Configures a maven project to produce a .jar archive with sources and javadocs.
 *
 * @experimental
 */
class MavenPackaging extends component_1.Component {
    /**
     * @experimental
     */
    constructor(project, pom, options = {}) {
        var _b, _c, _d;
        super(project);
        pom.addPlugin('org.apache.maven.plugins/maven-jar-plugin@3.2.0', {
            configuration: {
                archive: {
                    index: true,
                    manifest: {
                        addDefaultImplementationEntries: true,
                        addDefaultSpecificationEntries: true,
                    },
                },
            },
        });
        if ((_b = options.sources) !== null && _b !== void 0 ? _b : true) {
            pom.addPlugin('org.apache.maven.plugins/maven-source-plugin@3.2.1', {
                executions: [
                    { id: 'attach-sources', goals: ['jar'] },
                ],
            });
        }
        if ((_c = options.javadocs) !== null && _c !== void 0 ? _c : true) {
            pom.addPlugin('org.apache.maven.plugins/maven-javadoc-plugin@3.2.0', {
                executions: [
                    { id: 'attach-javadocs', goals: ['jar'] },
                ],
                configuration: {
                    failOnError: false,
                    show: 'protected',
                    sourceFileExcludes: { exclude: options.javadocsExclude },
                    detectJavaApiLink: false,
                    additionalJOptions: {
                        additionalJOption: [
                            '-J-XX:+TieredCompilation',
                            '-J-XX:TieredStopAtLevel=1',
                        ],
                    },
                },
            });
        }
        const env = {
            MAVEN_OPTS: '-XX:+TieredCompilation -XX:TieredStopAtLevel=1',
        };
        const distdir = (_d = options.distdir) !== null && _d !== void 0 ? _d : 'dist/java';
        this.task = project.addTask('package', {
            category: tasks_1.TaskCategory.RELEASE,
            description: `Creates a java deployment package under ${distdir}`,
            env,
        });
        this.task.exec(`mkdir -p ${distdir}`);
        this.task.exec(`mvn deploy -D=altDeploymentRepository=local::default::file:///$PWD/${distdir}`);
        project.gitignore.exclude(distdir);
    }
}
exports.MavenPackaging = MavenPackaging;
_a = JSII_RTTI_SYMBOL_1;
MavenPackaging[_a] = { fqn: "projen.java.MavenPackaging", version: "0.17.1" };
