// projen/lib/web/tailwind.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TailwindConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const json_1 = require("../json");
/**
 * (experimental) Declares a Tailwind CSS configuration file.
 *
 * There are multiple ways to add Tailwind CSS in your node project - see:
 * https://tailwindcss.com/docs/installation
 *
 * @see PostCss
 * @experimental
 */
class TailwindConfig {
    /**
     * @experimental
     */
    constructor(project, options) {
        var _b, _c;
        this.fileName = (_b = options === null || options === void 0 ? void 0 : options.fileName) !== null && _b !== void 0 ? _b : 'tailwind.config.json';
        this.file = new json_1.JsonFile(project, this.fileName, {
            obj: {
                purge: [],
                darkMode: false,
                theme: {
                    extend: {},
                },
                variants: {
                    extend: {},
                },
                plugins: [],
            },
        });
        (_c = project.npmignore) === null || _c === void 0 ? void 0 : _c.exclude(`/${this.fileName}`);
    }
}
exports.TailwindConfig = TailwindConfig;
_a = JSII_RTTI_SYMBOL_1;
TailwindConfig[_a] = { fqn: "projen.web.TailwindConfig", version: "0.17.1" };
