// projen/lib/python/setuppy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SetupPy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const file_1 = require("../file");
/**
 * (experimental) Python packaging script where package metadata can be placed.
 *
 * @experimental
 */
class SetupPy extends file_1.FileBase {
    /**
     * @experimental
     */
    constructor(project, options) {
        super(project, 'setup.py');
        this.setupConfig = {
            name: project.name,
            packages: [project.moduleName],
            python_requires: '>=3.6',
            classifiers: [
                'Intended Audience :: Developers',
                'Programming Language :: Python :: 3 :: Only',
                'Programming Language :: Python :: 3.6',
                'Programming Language :: Python :: 3.7',
                'Programming Language :: Python :: 3.8',
                'Programming Language :: Python :: 3.9',
            ],
            ...options ? this.renameFields(options) : [],
        };
    }
    /**
     * (experimental) Implemented by derived classes and returns the contents of the file to emit.
     *
     * @experimental
     */
    synthesizeContent(resolver) {
        const lines = [
            `# ${file_1.FileBase.PROJEN_MARKER}`,
            '',
            'import json',
            'from setuptools import setup',
            '',
            'kwargs = json.loads(',
            '    """',
            JSON.stringify(this.setupConfig, null, 4),
            '"""',
            ')',
            '',
            'setup(**kwargs)',
        ];
        return `${resolver.resolve(lines).join('\n')}\n`;
    }
    // modify some key names since JSII interfaces require fields to be camelCase
    renameFields(options) {
        const obj = {};
        for (const [key, value] of Object.entries(options)) {
            if (key === 'authorName') {
                obj.author = value;
            }
            else if (key === 'authorEmail') {
                obj.author_email = value;
            }
            else if (key === 'homepage') {
                obj.url = value;
            }
            else {
                obj[key] = value;
            }
        }
        return obj;
    }
}
exports.SetupPy = SetupPy;
_a = JSII_RTTI_SYMBOL_1;
SetupPy[_a] = { fqn: "projen.python.SetupPy", version: "0.17.1" };
