// projen/lib/python/setuptools.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Setuptools = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const component_1 = require("../component");
const tasks_1 = require("../tasks");
const setuppy_1 = require("./setuppy");
/**
 * (experimental) Manages packaging through setuptools with a setup.py script.
 *
 * @experimental
 */
class Setuptools extends component_1.Component {
    /**
     * @experimental
     */
    constructor(project, options) {
        super(project);
        project.addDevDependency('wheel@0.36.2');
        project.addDevDependency('twine@3.3.0');
        this.packageTask = project.addTask('package', {
            description: 'Creates source archive and wheel for distribution.',
            category: tasks_1.TaskCategory.RELEASE,
            exec: 'python setup.py sdist bdist_wheel',
        });
        this.publishTestTask = project.addTask('publish:test', {
            description: 'Uploads the package against a test PyPI endpoint.',
            category: tasks_1.TaskCategory.RELEASE,
            exec: 'twine upload --repository-url https://test.pypi.org/legacy/ dist/*',
        });
        this.publishTask = project.addTask('publish', {
            description: 'Uploads the package against a test PyPI endpoint.',
            category: tasks_1.TaskCategory.RELEASE,
            exec: 'twine upload dist/*',
        });
        new setuppy_1.SetupPy(project, {
            name: project.name,
            packages: [project.moduleName],
            authorName: options.authorName,
            authorEmail: options.authorEmail,
            version: options.version,
            description: options.description,
            license: options.license,
            homepage: options.homepage,
            classifiers: options.classifiers,
            ...options.setupConfig,
        });
    }
}
exports.Setuptools = Setuptools;
_a = JSII_RTTI_SYMBOL_1;
Setuptools[_a] = { fqn: "projen.python.Setuptools", version: "0.17.1" };
