// @serverless-stack/resources/dist/Topic.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Topic = void 0;
const cdk = __importStar(require("@aws-cdk/core"));
const sns = __importStar(require("@aws-cdk/aws-sns"));
const snsSubscriptions = __importStar(require("@aws-cdk/aws-sns-subscriptions"));
const Function_1 = require("./Function");
/////////////////////
// Construct
/////////////////////
class Topic extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const root = scope.node.root;
        const { 
        // Topic props
        snsTopic, 
        // Function props
        subscribers, } = props || {};
        this.subscriberFunctions = [];
        this.permissionsAttachedForAllSubscribers = [];
        ////////////////////
        // Create Topic
        ////////////////////
        if (cdk.Construct.isConstruct(snsTopic)) {
            this.snsTopic = snsTopic;
        }
        else {
            const snsTopicProps = (snsTopic || {});
            this.snsTopic = new sns.Topic(this, "Topic", Object.assign({ topicName: root.logicalPrefixedName(id) }, snsTopicProps));
        }
        ///////////////////////////
        // Create Subscribers
        ///////////////////////////
        this.addSubscribers(this, subscribers || []);
    }
    addSubscriber(scope, subscriber) {
        let fn;
        const i = this.subscriberFunctions.length;
        // subscriber is props
        if (subscriber.function) {
            subscriber = subscriber;
            fn = Function_1.Function.fromDefinition(scope, `Subscriber_${i}`, subscriber.function);
            this.snsTopic.addSubscription(new snsSubscriptions.LambdaSubscription(fn, subscriber.subscriberProps));
            this.subscriberFunctions.push(fn);
        }
        // subscriber is function
        else {
            subscriber = subscriber;
            fn = Function_1.Function.fromDefinition(scope, `Subscriber_${i}`, subscriber);
            this.snsTopic.addSubscription(new snsSubscriptions.LambdaSubscription(fn));
            this.subscriberFunctions.push(fn);
        }
        return fn;
    }
    addSubscribers(scope, subscribers) {
        subscribers.forEach((subscriber) => {
            // add subscriber
            const fn = this.addSubscriber(scope, subscriber);
            // attached existing permissions
            this.permissionsAttachedForAllSubscribers.forEach((permissions) => fn.attachPermissions(permissions));
        });
    }
    attachPermissions(permissions) {
        this.subscriberFunctions.forEach((subscriber) => subscriber.attachPermissions(permissions));
        this.permissionsAttachedForAllSubscribers.push(permissions);
    }
    attachPermissionsToSubscriber(index, permissions) {
        this.subscriberFunctions[index].attachPermissions(permissions);
    }
}
exports.Topic = Topic;
