// projen/lib/__tests__/jest.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("..");
const common_1 = require("../common");
const logging = require("../logging");
const util_1 = require("./util");
logging.disable();
const compilerOptionDefaults = {
    alwaysStrict: true,
    declaration: true,
    experimentalDecorators: true,
    inlineSourceMap: true,
    inlineSources: true,
    lib: ['es2018'],
    module: 'CommonJS',
    noEmitOnError: false,
    noFallthroughCasesInSwitch: true,
    noImplicitAny: true,
    noImplicitReturns: true,
    noImplicitThis: true,
    noUnusedLocals: true,
    noUnusedParameters: true,
    resolveJsonModule: true,
    strict: true,
    strictNullChecks: true,
    strictPropertyInitialization: true,
    stripInternal: true,
    target: 'ES2018',
};
test('Node Project Jest Defaults Configured', () => {
    var _a, _b, _c;
    const project = new __1.NodeProject({
        outdir: util_1.mkdtemp(),
        name: 'test-node-project',
        mergify: false,
        projenDevDependency: false,
        defaultReleaseBranch: 'master',
        jest: true,
        logging: {
            level: __1.LogLevel.OFF,
        },
    });
    expect((_a = project.jest) === null || _a === void 0 ? void 0 : _a.config).toBeTruthy();
    expect((_b = project.jest) === null || _b === void 0 ? void 0 : _b.config.clearMocks).toEqual(true);
    expect((_c = project.jest) === null || _c === void 0 ? void 0 : _c.config.collectCoverage).toEqual(true);
    const snapshot = util_1.synthSnapshot(project);
    expect(snapshot['package.json'].jest).toBeTruthy();
    const jest = snapshot['package.json'].jest;
    expect(jest.clearMocks).toEqual(true);
    expect(jest.collectCoverage).toEqual(true);
    expect(jest.coverageDirectory).toEqual('coverage');
});
test('Node Project Jest With Options Configured', () => {
    const project = new __1.NodeProject({
        outdir: util_1.mkdtemp(),
        name: 'test-node-project',
        defaultReleaseBranch: 'master',
        mergify: false,
        projenDevDependency: false,
        jest: true,
        jestOptions: {
            jestConfig: {
                automock: true,
                bail: 5,
                notify: false,
            },
        },
        logging: {
            level: __1.LogLevel.OFF,
        },
    });
    const snapshot = util_1.synthSnapshot(project);
    expect(snapshot['package.json'].jest).toBeTruthy();
    const jest = snapshot['package.json'].jest;
    expect(jest.automock).toEqual(true);
    expect(jest.bail).toEqual(5);
    expect(jest.notify).toEqual(false);
});
test('Typescript Project Jest Defaults Configured', () => {
    // WHEN
    const project = new __1.TypeScriptProject({
        outdir: util_1.mkdtemp(),
        name: 'test-typescript-project',
        defaultReleaseBranch: 'master',
        mergify: false,
        projenDevDependency: false,
        jest: true,
        logging: {
            level: __1.LogLevel.OFF,
        },
    });
    const snapshot = util_1.synthSnapshot(project);
    const jestTypescriptConfig = snapshot['tsconfig.jest.json'];
    expect(jestTypescriptConfig.compilerOptions).toBeTruthy();
    expect(jestTypescriptConfig.compilerOptions).toStrictEqual(compilerOptionDefaults);
    expect(jestTypescriptConfig.include).toEqual([common_1.PROJEN_RC, 'src/**/*.ts', 'test/**/*.ts']);
    expect(jestTypescriptConfig.exclude).toEqual(['node_modules']);
});
test('Typescript Project Jest With Compiler Options', () => {
    const compilerOptions = {
        esModuleInterop: true,
        noImplicitAny: false,
    };
    const project = new __1.TypeScriptProject({
        outdir: util_1.mkdtemp(),
        name: 'test-typescript-project',
        defaultReleaseBranch: 'master',
        mergify: false,
        projenDevDependency: false,
        jest: true,
        jestOptions: {
            typescriptConfig: {
                compilerOptions,
            },
        },
        logging: {
            level: __1.LogLevel.OFF,
        },
    });
    const mergedCompilerOptions = {
        ...compilerOptionDefaults,
        ...compilerOptions,
    };
    const snapshot = util_1.synthSnapshot(project);
    const jestTypescriptConfig = snapshot['tsconfig.jest.json'];
    expect(jestTypescriptConfig.compilerOptions).toBeTruthy();
    expect(jestTypescriptConfig.compilerOptions).toStrictEqual(mergedCompilerOptions);
});
test('testdir is under src', () => {
    // WHEN
    const project = new __1.TypeScriptProject({
        outdir: util_1.mkdtemp(),
        defaultReleaseBranch: 'master',
        logging: { level: __1.LogLevel.OFF },
        name: 'test-typescript-project',
        srcdir: 'mysrc',
        testdir: 'mysrc/boom/bam/__tests',
    });
    // THEN
    const files = util_1.synthSnapshot(project);
    expect(files['tsconfig.jest.json']).toBeUndefined(); // no special tsconfig for jest in this case
    expect(files['package.json'].jest.testMatch).toStrictEqual(['**/lib/boom/bam/__tests/**/?(*.)+(spec|test).js?(x)']);
});
test('addTestMatch() can be used to add patterns', () => {
    // GIVEN
    const project = new __1.NodeProject({
        outdir: util_1.mkdtemp(),
        defaultReleaseBranch: 'master',
        name: 'test',
        logging: {
            level: __1.LogLevel.OFF,
        },
    });
    const jest = new __1.Jest(project, { jestConfig: { testMatch: [] } });
    // WHEN
    jest.addTestMatch('foo/**');
    jest.addTestMatch('bar/baz/**');
    // THEN
    expect(util_1.synthSnapshot(project)['package.json'].jest.testMatch).toStrictEqual([
        'foo/**',
        'bar/baz/**',
    ]);
});
