// projen/lib/__tests__/json.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("..");
const util_1 = require("./util");
test('json object can be mutated before synthesis', () => {
    const prj = new util_1.TestProject();
    const obj = {
        hello: 'world',
    };
    new __1.JsonFile(prj, 'my/json/file.json', { obj });
    // mutate obj (should be reflected in the output)
    obj.anotherField = {
        foo: 1234,
    };
    expect(util_1.synthSnapshot(prj)['my/json/file.json']).toStrictEqual({
        hello: 'world',
        anotherField: { foo: 1234 },
    });
});
test('omitEmpty', () => {
    // GIVEN
    const p = new util_1.TestProject();
    // WHEN
    new __1.JsonFile(p, 'file.json', {
        omitEmpty: true,
        obj: {
            hello: 1234,
            empty: {},
            array_with_undefined: [undefined, 123, 456],
            child: {
                with: 'hello',
                empty: {
                    subchild: {},
                },
                empty_strings_are_not_omitted: '',
                zeros_are_not_omitted: 0,
                empty_array: [],
                array_with_empty_objects: [{}, 123],
                array_with_empty_subobjects: [
                    {
                        i_am_empty: {},
                        i_am_not: 'hi there',
                    },
                    {
                        just: 1234,
                    },
                ],
            },
        },
    });
    // THEN
    expect(util_1.synthSnapshot(p)['file.json']).toStrictEqual({
        hello: 1234,
        array_with_undefined: [123, 456],
        child: {
            with: 'hello',
            empty_strings_are_not_omitted: '',
            zeros_are_not_omitted: 0,
            array_with_empty_objects: [123],
            array_with_empty_subobjects: [
                {
                    i_am_not: 'hi there',
                },
                {
                    just: 1234,
                },
            ],
        },
    });
});
test('json file can contain projen marker', () => {
    const prj = new util_1.TestProject();
    const obj = {};
    new __1.JsonFile(prj, 'my/json/file-marker.json', { obj, marker: true });
    const output = util_1.synthSnapshot(prj)['my/json/file-marker.json'];
    expect(output['//']).toBe(__1.JsonFile.PROJEN_MARKER);
});
