// projen/lib/__tests__/makefile.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const __1 = require("..");
const util_1 = require("./util");
test('makefile synthesises correctly', () => {
    const prj = new util_1.TestProject();
    new __1.Makefile(prj, 'Makefile', {
        all: ['one', 'two', 'three'],
        rules: [
            {
                targets: ['one'],
                recipe: ['touch one'],
                phony: false,
            },
            {
                targets: ['two', 'three'],
                prerequisites: ['one'],
                recipe: ['touch $@'],
            },
            {
                targets: ['clean'],
                recipe: ['rm -f one two three'],
                phony: true,
            },
        ],
    });
    expect(util_1.synthSnapshot(prj).Makefile).toStrictEqual([
        '.PHONY: all',
        'all: one two three',
        '',
        'one:',
        '\ttouch one',
        '',
        'two three: one',
        '\ttouch $@',
        '',
        '.PHONY: clean',
        'clean:',
        '\trm -f one two three',
        '',
    ].join('\n'));
});
test('makefile synthesises correctly using imperative API', () => {
    const prj = new util_1.TestProject();
    new __1.Makefile(prj, 'Makefile')
        .addRule({
        targets: ['one'],
        recipe: ['touch one'],
        phony: false,
    })
        .addRules({
        targets: ['two', 'three'],
        prerequisites: ['one'],
        recipe: ['touch $@'],
    }, {
        targets: ['clean'],
        recipe: ['rm -f one two three'],
        phony: true,
    })
        .addAll('one')
        .addAlls('two', 'three');
    expect(util_1.synthSnapshot(prj).Makefile).toStrictEqual([
        '.PHONY: all',
        'all: one two three',
        '',
        'one:',
        '\ttouch one',
        '',
        'two three: one',
        '\ttouch $@',
        '',
        '.PHONY: clean',
        'clean:',
        '\trm -f one two three',
        '',
    ].join('\n'));
});
