// projen/lib/__tests__/resolve.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _resolve_1 = require("../_resolve");
test('null', () => {
    expect(_resolve_1.resolve(null)).toStrictEqual(null);
});
test('undefined', () => {
    expect(_resolve_1.resolve(undefined)).toStrictEqual(undefined);
});
test('string', () => {
    expect(_resolve_1.resolve('')).toStrictEqual('');
    expect(_resolve_1.resolve('hello')).toStrictEqual('hello');
    expect(_resolve_1.resolve('0')).toStrictEqual('0');
    expect(_resolve_1.resolve('hello world')).toStrictEqual('hello world');
    expect(_resolve_1.resolve('  ')).toStrictEqual('  ');
});
test('number', () => {
    expect(_resolve_1.resolve(0)).toStrictEqual(0);
    expect(_resolve_1.resolve(10)).toStrictEqual(10);
    expect(_resolve_1.resolve(-102)).toStrictEqual(-102);
});
test('boolean', () => {
    expect(_resolve_1.resolve(true)).toStrictEqual(true);
    expect(_resolve_1.resolve(false)).toStrictEqual(false);
});
test('array', () => {
    expect(_resolve_1.resolve([])).toStrictEqual([]);
    expect(_resolve_1.resolve([1, 2, 3])).toStrictEqual([1, 2, 3]);
});
test('object', () => {
    expect(_resolve_1.resolve({})).toStrictEqual({});
    expect(_resolve_1.resolve({ foo: 123, hello: [1, 2, { bar: 3 }] })).toStrictEqual({
        foo: 123,
        hello: [1, 2, { bar: 3 }],
    });
});
test('functions are resolved', () => {
    expect(_resolve_1.resolve(() => 123)).toStrictEqual(123);
    expect(_resolve_1.resolve(() => 'hello')).toStrictEqual('hello');
    expect(_resolve_1.resolve(() => undefined)).toStrictEqual(undefined);
});
test('recursive resolve', () => {
    expect(_resolve_1.resolve(() => [1, 2, () => 30])).toStrictEqual([1, 2, 30]);
    expect(_resolve_1.resolve(() => ({ foo: 123, bar: () => 'bar' }))).toStrictEqual({ foo: 123, bar: 'bar' });
    expect(_resolve_1.resolve(() => ({ foo: 123, bar: () => ['bar', 'baz', { hello: () => 'world' }] }))).toStrictEqual({
        foo: 123,
        bar: ['bar', 'baz', { hello: 'world' }],
    });
});
test('context is passed to functions', () => {
    expect(_resolve_1.resolve((x) => x + 1, { args: [10] })).toStrictEqual(11);
    const calculator = {
        add: (a, b) => a + b,
        sub: (a, b) => a - b,
        mul: (a, b) => a * b,
        div: (a, b) => a / b,
        fun: (a1, b1, text) => [
            `this is ${text}`,
            (a2) => a1 * a2,
            (_, b2) => b1 + b2,
            {
                another: {
                    level: (a3, b3, text3) => `hello ${text3} ${b3} ${a3}`,
                },
            },
        ],
    };
    expect(_resolve_1.resolve(calculator, { args: [10, 2, 'fun'] })).toStrictEqual({
        add: 12,
        sub: 8,
        mul: 20,
        div: 5,
        fun: [
            'this is fun',
            100,
            4,
            { another: { level: 'hello fun 2 10' } },
        ],
    });
});
test('"undefined" values are omitted', () => {
    const r = (o) => _resolve_1.resolve(o);
    expect(r({ foo: undefined })).toStrictEqual({});
    expect(r({ foo: { bar: undefined } })).toStrictEqual({ foo: {} });
    expect(r({ foo: { bar: undefined, hello: 123 } })).toStrictEqual({ foo: { hello: 123 } });
    expect(r({ foo: [undefined] })).toStrictEqual({ foo: [] });
});
test('omitEmpty', () => {
    const r = (o) => _resolve_1.resolve(o, { omitEmpty: true });
    expect(r({ foo: {} })).toStrictEqual(undefined);
    expect(r({ foo: { bar: {} } })).toStrictEqual(undefined);
    expect(r({ foo: [] })).toStrictEqual(undefined);
    expect(r({ bar: { foo: [] } })).toStrictEqual(undefined);
    expect(r({ foo: { hello: 'world', boom: { bar: [] } } })).toStrictEqual({ foo: { hello: 'world' } });
});
